File: packages/autonavi_maps_flutter/android/src/main/kotlin/com/autonavi/maps/flutter/AMapController.kt
package com.autonavi.maps.flutter

import android.content.Context
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

class AMapController(
    context: Context,
    messenger: BinaryMessenger,
    viewId: Int,
    params: Map<String, Any>,
) : PlatformView, MethodChannel.MethodCallHandler {

    private val adapter = AMapSDKAdapter(context)
    private val channel = MethodChannel(messenger, "plugins.autonavi.flutter/amap_map_$viewId")

    init {
        channel.setMethodCallHandler(this)

        // Wire adapter event callbacks → Flutter channel
        adapter.onCameraMove   = { args -> channel.invokeMethod("camera#onMove", args) }
        adapter.onCameraIdle   = { channel.invokeMethod("camera#onIdle", null) }
        adapter.onMapTap       = { args -> channel.invokeMethod("map#onTap", args) }
        adapter.onMapLongPress = { args -> channel.invokeMethod("map#onLongPress", args) }
        adapter.onMarkerTap    = { id -> channel.invokeMethod("marker#onTap", mapOf("markerId" to id)) }

        // Initial state
        @Suppress("UNCHECKED_CAST")
        (params["initialCameraPosition"] as? Map<String, Any>)?.let { adapter.applyInitialCamera(it) }

        @Suppress("UNCHECKED_CAST")
        (params["options"] as? Map<String, Any>)?.let { adapter.applyOptions(it) }

        @Suppress("UNCHECKED_CAST")
        (params["markersToAdd"] as? List<Map<String, Any>>)?.forEach { adapter.addMarker(it) }

        @Suppress("UNCHECKED_CAST")
        (params["polylinesToAdd"] as? List<Map<*, *>>)?.let {
            adapter.handlePolylineUpdates(mapOf("polylinesToAdd" to it))
        }

        @Suppress("UNCHECKED_CAST")
        (params["polygonsToAdd"] as? List<Map<*, *>>)?.let {
            adapter.handlePolygonUpdates(mapOf("polygonsToAdd" to it))
        }

        @Suppress("UNCHECKED_CAST")
        (params["circlesToAdd"] as? List<Map<*, *>>)?.let {
            adapter.handleCircleUpdates(mapOf("circlesToAdd" to it))
        }
    }

    override fun getView(): View = adapter.nativeView()

    override fun dispose() {
        channel.setMethodCallHandler(null)
        adapter.destroy()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "map#update" -> {
                @Suppress("UNCHECKED_CAST")
                call.argument<Map<String, Any>>("options")?.let { adapter.applyOptions(it) }
                result.success(null)
            }
            "map#moveCamera" -> {
                @Suppress("UNCHECKED_CAST")
                adapter.applyCameraUpdate(call.arguments as Map<*, *>, animated = false)
                result.success(null)
            }
            "map#animateCamera" -> {
                @Suppress("UNCHECKED_CAST")
                adapter.applyCameraUpdate(call.arguments as Map<*, *>, animated = true)
                result.success(null)
            }
            "map#getCameraPosition" -> result.success(adapter.currentCameraPosition())
            "map#getLatLng" -> {
                @Suppress("UNCHECKED_CAST")
                val args = call.arguments as Map<*, *>
                result.success(adapter.latLng(
                    fromScreen = (args["x"] as Number).toInt(),
                    y = (args["y"] as Number).toInt()
                ))
            }
            "map#getScreenCoordinate" -> {
                @Suppress("UNCHECKED_CAST")
                result.success(adapter.screenCoordinate(forLatLng = call.arguments as Map<*, *>))
            }
            "map#takeSnapshot" -> {
                adapter.takeSnapshot { bytes -> result.success(bytes) }
            }
            "markers#update" -> {
                @Suppress("UNCHECKED_CAST")
                adapter.handleMarkerUpdates(call.arguments as Map<*, *>)
                result.success(null)
            }
            "markers#showInfoWindow" -> {
                adapter.showInfoWindow(call.argument<String>("markerId") ?: "")
                result.success(null)
            }
            "markers#hideInfoWindow" -> {
                adapter.hideInfoWindow(call.argument<String>("markerId") ?: "")
                result.success(null)
            }
            "markers#isInfoWindowShown" -> {
                result.success(adapter.isInfoWindowShown(call.argument<String>("markerId") ?: ""))
            }
            "polylines#update" -> {
                @Suppress("UNCHECKED_CAST")
                adapter.handlePolylineUpdates(call.arguments as Map<*, *>)
                result.success(null)
            }
            "polygons#update" -> {
                @Suppress("UNCHECKED_CAST")
                adapter.handlePolygonUpdates(call.arguments as Map<*, *>)
                result.success(null)
            }
            "circles#update" -> {
                @Suppress("UNCHECKED_CAST")
                adapter.handleCircleUpdates(call.arguments as Map<*, *>)
                result.success(null)
            }
            "coordinate#convertFromWGS84" -> {
                @Suppress("UNCHECKED_CAST")
                result.success(adapter.convertFromWGS84(call.arguments as Map<*, *>))
            }
            else -> result.notImplemented()
        }
    }
}

File: packages/autonavi_maps_flutter/ios/Classes/AMapController.swift
import Flutter

class AMapController: NSObject, FlutterPlatformView {

    private let adapter: AMapSDKAdapter
    private let channel: FlutterMethodChannel

    init(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        params: [String: Any]
    ) {
        adapter = AMapSDKAdapter(frame: frame)
        channel = FlutterMethodChannel(
            name: "plugins.autonavi.flutter/amap_map_\(viewId)",
            binaryMessenger: messenger
        )
        super.init()

        // Wire adapter event callbacks → Flutter channel
        adapter.onCameraMove   = { [weak self] args in self?.channel.invokeMethod("camera#onMove", arguments: args) }
        adapter.onCameraIdle   = { [weak self] in self?.channel.invokeMethod("camera#onIdle", arguments: nil) }
        adapter.onMapTap       = { [weak self] args in self?.channel.invokeMethod("map#onTap", arguments: args) }
        adapter.onMapLongPress = { [weak self] args in self?.channel.invokeMethod("map#onLongPress", arguments: args) }
        adapter.onMarkerTap    = { [weak self] id in self?.channel.invokeMethod("marker#onTap", arguments: ["markerId": id]) }

        // Initial state
        if let cameraJson = params["initialCameraPosition"] as? [AnyHashable: Any] {
            adapter.applyInitialCamera(cameraJson)
        }
        if let options = params["options"] as? [String: Any] {
            adapter.applyOptions(options)
        }
        if let markersToAdd = params["markersToAdd"] as? [[AnyHashable: Any]] {
            markersToAdd.forEach { adapter.addMarker($0) }
        }
        if let polylinesToAdd = params["polylinesToAdd"] as? [[AnyHashable: Any]] {
            adapter.handlePolylineUpdates(["polylinesToAdd": polylinesToAdd])
        }
        if let polygonsToAdd = params["polygonsToAdd"] as? [[AnyHashable: Any]] {
            adapter.handlePolygonUpdates(["polygonsToAdd": polygonsToAdd])
        }
        if let circlesToAdd = params["circlesToAdd"] as? [[AnyHashable: Any]] {
            adapter.handleCircleUpdates(["circlesToAdd": circlesToAdd])
        }

        channel.setMethodCallHandler(onMethodCall)
    }

    func view() -> UIView { adapter.nativeView() }

    // MARK: - Method channel

    private func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments

        switch call.method {
        case "map#update":
            if let options = (args as? [String: Any])?["options"] as? [String: Any] {
                adapter.applyOptions(options)
            }
            result(nil)

        case "map#moveCamera":
            if let json = args as? [AnyHashable: Any] { adapter.applyCameraUpdate(json, animated: false) }
            result(nil)

        case "map#animateCamera":
            if let json = args as? [AnyHashable: Any] { adapter.applyCameraUpdate(json, animated: true) }
            result(nil)

        case "map#getCameraPosition":
            result(adapter.currentCameraPosition())

        case "map#getLatLng":
            if let json = args as? [String: Any],
               let x = json["x"] as? Int, let y = json["y"] as? Int {
                result(adapter.latLng(fromScreen: x, y: y))
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: nil, details: nil))
            }

        case "map#getScreenCoordinate":
            if let json = args as? [AnyHashable: Any] {
                result(adapter.screenCoordinate(forLatLng: json))
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: nil, details: nil))
            }

        case "map#takeSnapshot":
            result(adapter.takeSnapshot())

        case "markers#update":
            if let json = args as? [AnyHashable: Any] { adapter.handleMarkerUpdates(json) }
            result(nil)

        case "markers#showInfoWindow":
            if let id = (args as? [String: Any])?["markerId"] as? String { adapter.showInfoWindow(markerId: id) }
            result(nil)

        case "markers#hideInfoWindow":
            if let id = (args as? [String: Any])?["markerId"] as? String { adapter.hideInfoWindow(markerId: id) }
            result(nil)

        case "markers#isInfoWindowShown":
            let id = (args as? [String: Any])?["markerId"] as? String ?? ""
            result(adapter.isInfoWindowShown(markerId: id))

        case "polylines#update":
            if let json = args as? [AnyHashable: Any] { adapter.handlePolylineUpdates(json) }
            result(nil)

        case "polygons#update":
            if let json = args as? [AnyHashable: Any] { adapter.handlePolygonUpdates(json) }
            result(nil)

        case "circles#update":
            if let json = args as? [AnyHashable: Any] { adapter.handleCircleUpdates(json) }
            result(nil)

        case "coordinate#convertFromWGS84":
            if let json = args as? [AnyHashable: Any] {
                result(adapter.convertFromWGS84(json))
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: nil, details: nil))
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
